HighAvailabilityModule.Algorithm/MembershipWithWitness.cs

```diff
@@ -143,32 +147,66 @@ internal async Task HeartBeatAsPrimaryAsync()
         internal async Task KeepPrimaryAsync()
         {
             var token = this.AlgorithmCancellationToken;
-            while (this.RunningAsPrimary(DateTime.UtcNow))
+            while (this.RunningAsPrimary(DateTime.UtcNow) && AffiliatedASPrimary(DateTime.UtcNow))
             {
                 token.ThrowIfCancellationRequested();
 #pragma warning disable 4014
                 this.HeartBeatAsPrimaryAsync();
                 this.CheckPrimaryAsync(DateTime.UtcNow);
+                this.CheckAffiliatedAsync(DateTime.UtcNow);
 #pragma warning restore 4014
                 await Task.Delay(this.HeartBeatInterval, token).ConfigureAwait(false);
             }
-
             Trace.TraceWarning($"[{DateTime.UtcNow:O}][Protocol][{this.Uuid}] Lost Primary");
         }
 
         private bool PrimaryUp => this.lastSeenHeartBeat != default && !this.lastSeenHeartBeat.Entry.IsEmpty;
 
+        private bool AffiliatedPrimaryUp => this.lastSeenAffiliated != default && !this.lastSeenAffiliated.Entry.IsEmpty;
+
         internal bool RunningAsPrimary(DateTime now)
         {
             var primary = this.PrimaryUp && this.lastSeenHeartBeat.Entry.Uuid == this.Uuid && now - this.lastSeenHeartBeat.QueryTime < (this.HeartBeatTimeout - this.HeartBeatInterval);
             if (!primary)
             {
                 Trace.TraceInformation($"[{DateTime.UtcNow:O}][Protocol] Running as secondary {this.Dump()}.");
             }
-
             return primary;
         }
 
+        internal bool AffiliatedASPrimary(DateTime now)
+        {
+            var affiliatedPrimary = this.AffiliatedType == string.Empty || (this.AffiliatedPrimaryUp && this.lastSeenAffiliated.Entry.Uname.ToLower() == this.Uname.ToLower() && now - this.lastSeenAffiliated.QueryTime < (this.HeartBeatTimeout - this.HeartBeatInterval));
+            if (!affiliatedPrimary)
+            {
+                Trace.TraceInformation($"[{DateTime.UtcNow:O}][Protocol] Affiliated service is running on another machine. {this.Dump()}.");
+            }
+            return affiliatedPrimary;
+        }
+
+        internal async Task CheckAffiliatedAsync(DateTime now)
```

Comment:
This is basically the same as `HeartBeatAsPrimaryAsync`? Can we consolidate the logic?